### PR TITLE
feat: introduce digitization

### PR DESCRIPTION
**Pull request by asalzburger**
This PR superseded #652 and introduces the geometric digitization + the possibility to configure it with json.

In detail:
- This PR introduces a new configuration based on Json for digitization 
- It adds a new unit test `Examples/JsonDigitizationConfigTests.cpp`
- It introduces the new DigitizationAlgorithm capable of smearing and geometric digitization
- It allows to use the same `RootDigitizationWriter` for SmearingAlgorithm and DigitizationAlgorithm

When deployed, the `HitSmearing` and `PlanarSteppingAlgorithm` can become osbolete, and finally the `Plugins/Digitization` and `Plugins/Identification` can vanish.

Here are two example files snippes for the Pixel detector of the TrackML detector

1) analog digitization (blue)

```json
{
  "acts-geometry-hierarchy-map": {
    "format-version": 0,
    "value-identifier": "digitization-configuration"
  },
  "entries": [
    {
      "layer": 2,
      "value": {
        "geometric": {
          "digital": false,
          "indices": [
            0,
            1
          ],
          "segmentation": {
            "binningdata": [
              {
                "bins": 336,
                "max": 8.399999618530273,
                "min": -8.399999618530273,
                "option": "open",
                "type": "equidistant",
                "value": "binX"
              },
              {
                "bins": 1280,
                "max": 36.0,
                "min": -36.0,
                "option": "open",
                "type": "equidistant",
                "value": "binY"
              }
            ]
          },
          "thickness": 0.15,
          "threshold": 0.01
        }
      },
      "volume": 8
    }]
}
```

2) digital digitization (red)

(only showing the change)

```json 
          "digital": false,
```

The results can be seen in the following plot:
Screen Shot 2021-01-29 at 12 44 00

### CODE REVIEW

**Comment by asalzburger** [comment]
Hi @XiaocongAi - sorry for the bit lengthy Pr, but it should be quite clear. There will be a bit more code to come for the covariance assignment and the smearing generation of charge and lorentz vector, but this can be done as an addon.

**Comment by asalzburger** [comment]
@paulgessinger - I tried three times now with the `macos` build on this PR but it constantly fails :-(

**Comment by asalzburger** [comment]
Ok, CI finally runs through - @Xai - this one is ready for review

File: Examples/Algorithms/Digitization/src/DigitizationAlgorithm.cpp [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        [&](const auto& digitizer) {
+          for (auto h = moduleSimHits.begin(); h != moduleSimHits.end(); ++h) {
+            const auto& simHit = *h;
+            const auto simHitIdx = simHits.index_of(h);
+
+            DigitizedParameters dParameters;
+
+            // Geometric part - 0, 1, 2 local parameters are possible
+            if (not digitizer.geometric.indices.empty()) {
+              ACTS_VERBOSE("Configured to geometric digitize "
```
Comment (XiaocongAi) [inline Examples/Algorithms/Digitization/src/DigitizationAlgorithm.cpp:167]:
```suggestion
              ACTS_VERBOSE("Configured to geometrically digitize "
```

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    std::function<std::vector<Acts::ActsScalar>(size_t, size_t, RandomEngine&)>;
+
+// A void drift generator - no lorentz drift.
+struct VoidDriftGenerator {
+  /// Charge generation call operator
+  ///
+  /// @param pos is the position of the call (unused here)
+  /// @param rng is the random engine (unused here)
+  ///
+  /// @return the unchanged path length
```
Comment (XiaocongAi) [inline Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp:36]:
```suggestion
  /// @return the charge drift
```

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+using DriftGenerator =
+    std::function<Acts::Vector3(const Acts::Vector3&, RandomEngine&)>;
+using ChargeGenerator = std::function<Acts::ActsScalar(
+    Acts::ActsScalar, Acts::ActsScalar, RandomEngine&)>;
+using CovarianceGenerator =
+    std::function<std::vector<Acts::ActsScalar>(size_t, size_t, RandomEngine&)>;
+
+// A void drift generator - no lorentz drift.
+struct VoidDriftGenerator {
+  /// Charge generation call operator
```
Comment (XiaocongAi) [inline Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp:31]:
```suggestion
  /// Drift generation call operator
```

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  Acts::ActsScalar operator()(Acts::ActsScalar path,
+                              Acts::ActsScalar /*dlength, unused*/,
+                              RandomEngine& /*rng, unused*/) const {
+    return path;
+  }
+};
+
+/// A void covariance generator - returns empty vector
+struct VoidCovarianceGenerator {
+  /// Charge generation call operator
```
Comment (XiaocongAi) [inline Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp:61]:
```suggestion
  /// Covariance generation call operator
```

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  ///
+  /// @param b0 is the size of the cluster in first direction
+  /// @param b1 is the size of the cluster in second direction
+  /// @param rng is the random engine (unused here)
+  ///
+  /// @return the unchanged path length
```
Comment (XiaocongAi) [inline Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp:67]:
```suggestion
  /// @return the empty covariance
```

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+struct GeometricDigitizationConfig {
+  std::vector<Acts::BoundIndices> indices = {};
+  Acts::BinUtility segmentation;
+  /// Drift generation
+  DriftGenerator drift = VoidDriftGenerator{};
+  double thickness = 0.;
+  /// Charge generation
+  ChargeGenerator charge = VoidChargeGenerator{};
+  double threshold = 0.;
+  /// Position and Covariance generation
```
Comment (XiaocongAi) [inline Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp:91]: Is the `covariance` supposed to include both the position and covariance, or just covariance?

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationAlgorithm.hpp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+  /// @param ctx is the algorithm context with event information
+  /// @return a process code indication success or failure
+  ProcessCode execute(const AlgorithmContext& ctx) const final override;
+
+ private:
+  /// Nested struct for digitized parameters
+  struct DigitizedParameters {
+    std::vector<Acts::BoundIndices> indices = {};
+    std::vector<Acts::ActsScalar> values = {};
+    std::vector<Acts::ActsScalar> covariances = {};
```
Comment (XiaocongAi) [inline Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationAlgorithm.hpp:76]: I guess this vector is for just the errors of the parameters (not their correlation), right? Personally, I would expect a matrix from the name `covariance`. Is it better to rename it to e.g. `errors` or `uncertainties`?

File: Examples/Io/Json/src/JsonDigitizationConfig.cpp [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    nlohmann::json& j, const ActsExamples::GeometricDigitizationConfig& gdc) {
+  std::vector<size_t> indices;
+  for (const auto& idx : gdc.indices) {
+    indices.push_back(static_cast<size_t>(idx));
+  }
+  j["indices"] = indices;
+  j["segmentation"] = nlohmann::json(gdc.segmentation);
+  j["thickness"] = gdc.thickness;
+  j["threshold"] = gdc.threshold;
+  j["digital"] = gdc.digital;
```
Comment (XiaocongAi) [inline Examples/Io/Json/src/JsonDigitizationConfig.cpp:100]: No set for the `covariance`?

**Review by XiaocongAi — COMMENTED** [review]
Hi @asalzburger, I have not looked  into the Json part in great detail, but in general, the PR looks very nice to me. I have only a few minor comments.

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationAlgorithm.hpp [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+      indices[ei] = dParams.indices[ei];
+      par[ei] = dParams.values[ei];
+      cov(ei, ei) = dParams.covariances[ei];
+    }
+    return {indices, par, cov};
+  }
+
+  /// Construct a fixed-size smearer from a configuration.
+  ///
+  /// It's templated on the smearing dimention given by @tparam kSmearDim
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationAlgorithm.hpp:164]: Fixed.

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationAlgorithm.hpp [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+  /// @param isl The indexed source link for the measurement
+  ///
+  /// @return a variant measurement
+  Measurement createMeasurement(const DigitizedParameters& dParams,
+                                const IndexSourceLink& isl) const
+      noexcept(false);
+
+  /// Nested smearer struct that holds geometric digitizer and smearing
+  /// Support up to 4 dimensions.
+  template <size_t kGeoDIM, size_t kSmearDIM = 0>
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationAlgorithm.hpp:119]: That was actually buggy ... oh oh, fixed it.

File: Examples/Algorithms/Digitization/src/DigitizationAlgorithm.cpp (hunk 1 above)
Comment (asalzburger) [reply Examples/Algorithms/Digitization/src/DigitizationAlgorithm.cpp:167]: Fixed.

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp (hunk 2 above)
Comment (asalzburger) [reply Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp:36]: Copy & past - fixed.

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp (hunk 3 above)
Comment (asalzburger) [reply Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp:31]: Fixed.

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp (hunk 4 above)
Comment (asalzburger) [reply Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp:61]: Also fixed.

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp (hunk 6 above)
Comment (asalzburger) [reply Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationConfig.hpp:91]: Just covariance.

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationAlgorithm.hpp (hunk 7 above)
Comment (asalzburger) [reply Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationAlgorithm.hpp:76]: Measurement directions are by all means independent for our purpose.
We could write 'variances' ?

File: Examples/Io/Json/src/JsonDigitizationConfig.cpp (hunk 8 above)
Comment (asalzburger) [reply Examples/Io/Json/src/JsonDigitizationConfig.cpp:100]: That needs a follow-up PR with fillers.

**Comment by asalzburger** [comment]
@XiaocongAi - I think I have addressed everything & simplified the code by removing the 'Void...Generators' by just having void functions.

**Review by XiaocongAi — APPROVED** [review]
HI @asalzburger , thank you for the changes. I will approve.
One thing to confirm though, it looks to me we don't need the `SmearingAlgorithm` since it's basically contained in this `DigitizationAlgorithm` when we use the `smearing` instead of `geometric` in the Json input. But you mentioned the `HitSmearing`. Which one are we going to remove, the `SmearingAlgorithm` or the `HitSmearing`?

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationAlgorithm.hpp (hunk 7 above)
Comment (XiaocongAi) [reply Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationAlgorithm.hpp]: Sounds good!

**Comment by asalzburger** [comment]
Ideally, both, 'HitSmearing' and 'SmearingAlgorithm' could go, but I would like to keep the latter as we can configure it through command options.

**Pull request merged**